server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/RepositoryUtil.java

```diff
@@ -113,14 +116,32 @@ private static String concatenatePaths(Iterable<MergeSource> mergeSources) {
         return Streams.stream(mergeSources).map(MergeSource::path).collect(Collectors.joining(","));
     }
 
-    public static <T> void completeQueryExecutionException(CompletableFuture<MergedEntry<T>> future,
-                                                           Throwable cause) {
-        if (cause instanceof CentralDogmaException) {
-            future.completeExceptionally(cause);
-            return;
+    /**
+     * Applies the specified {@link Query} to the {@link Entry#content()} of the specified {@link Entry} and
+     * returns the query result.
+     *
+     * @throws IllegalStateException if the specified {@link Entry} is a directory
+     * @throws QuerySyntaxException if the syntax of specified {@link Query} is invalid
+     * @throws QueryExecutionException if an {@link Exception} is raised while applying the specified
+     *                                 {@link Query} to the {@link Entry#content()}
+     */
+    public static <T> Entry<T> applyQuery(Entry<T> entry, Query<T> query) {
+        requireNonNull(query, "query");
+        entry.content(); // Ensure that content is not null.
+        final EntryType entryType = entry.type();
+
+        final QueryType queryType = query.type();
+        if (!queryType.supportedEntryTypes().contains(entryType)) {
+            throw new QueryExecutionException(query + " does not support this entry type: " + entryType);
         }
 
-        future.completeExceptionally(new QueryExecutionException(cause));
+        if (queryType == IDENTITY) {
+            return entry;
+        } else if (queryType == JSON_PATH) {
+            return Entry.of(entry.revision(), query.path(), entryType, query.apply(entry.content()));
+        } else {
+            throw new QueryExecutionException(query + " does not support this entry type: " + entryType);
```

Comment:
How about:

`"Unsupported entry type: " + entryType + " (query: " + query + ')'`

because the query string may be pretty long.